Cross compile for Linux, MacOS and Windows on CI

owenthereal opened the pull request:
This PR includes the following changes:

* Cross compile for Linux, MacOS and Windows on CI and upload corresponding artifacts to a release ([example](https://github.com/owenthereal/jq/releases/tag/jq-0.0.22))
* Remove `ubuntu-20.04` build because
  * [A static bin couldn't be built with the gcc toolchain](https://github.com/jqlang/jq/pull/2620#discussion_r1249458361)
  * [The aarch64 clang build failed for weird reasons](https://github.com/owenthereal/jq/actions/runs/5483326592/jobs/9989546559#step:4:428)
* Remove `gcc` compile target on MacOS because it symlinks `gcc` to `clang`
* Remove compiler variant for Windows because only `gcc` is supported
* Speed up `release` job by not waiting on `docker` job because it doesn't require anything from the `docker` job

This closes https://github.com/jqlang/jq/pull/2618 https://github.com/jqlang/jq/issues/2386

cc: @wader @itchyny

owenthereal committed:
Support arm64 build for MacOS

owenthereal committed:
Support aarch64 build for Linux

owenthereal committed:
Properly cross compile for MacOS

owenthereal committed:
Remove compiler matrix from Windows

Only `gcc` is supported well for msys2.

owenthereal committed:
Make release job not wait for docker job

The CI release job doesn't need anything from the Docker job

wader commented on .github/workflows/ci.yml:
```diff
@@ -48,10 +49,15 @@ jobs:
             --disable-valgrind \
             --with-oniguruma=builtin \
-            ${{ matrix.configure_flag }}
+            --host=${{ matrix.arch }}-linux-gnu \
```
I started looking into doing cross builds yesterday but ended up reading old autoconf documentation about host, target and build and got irritated 😄 what a mess. Also a bit confused why jq's configure script only has `--build` and `--host`... and `--host` is what i would call "target"? :) it's generated by autoconf 2.71 which is quite modern.

Anyways thanks doing this!

nicowilliams replied:
@wader https://github.com/jqlang/jq/wiki/Cross-compilation

owenthereal replied:
I added cross-compilation for Windows as well. However, I noticed a few things:

* Tests with the `mingw-w64-x86_64` build [failed](https://github.com/owenthereal/jq/actions/runs/5498251723/jobs/10019578971) for segmentation fault:  (Maybe the same issue as the [clang issue](https://discord.com/channels/1108476965947723846/1119871041997652038/1122924762075758723)). I fallback to using msys gcc for x86_64 build
* For Windows i386 build, I needed to add the `__MINGW32__` guard to `include <pthread.h>` instead of the using `pthread_*` functions defined in `jv.c` & `jv_thread.h` due to "multiple definition" of the same functions ([Ref](https://github.com/owenthereal/jq/actions/runs/5496335434/jobs/10016293755#step:4:1237)).

wader commented on .github/workflows/ci.yml:
```diff
@@ -48,10 +49,15 @@ jobs:
           make
-          strip jq
+          ${ARCH}-linux-gnu-strip jq
+          file ./jq
```
Accidentally left in? i think it could make sense to keep

nicowilliams replied:
It's debugging, but it doesn't hurt.

owenthereal replied:
The `file` line is to ensure the cross-compiled binary is what we expect. I guess it could be removed but it doesn't hurt to have that visibility.

wader commented on .github/workflows/ci.yml:
```diff
@@ -92,6 +97,9 @@ jobs:
           brew install autoconf automake libtool
+      - name: "Set CC"
+        run: |
+          echo "CC=clang -target ${{ matrix.arch }}-apple-darwin$(uname -r)" >> $GITHUB_ENV
```
Can't use `env:` because of uname i guess?

owenthereal replied:
Yup, that's the exact reason. The value of `env` is a string instead of being evaluated.

wader commented on .github/workflows/ci.yml:
```diff
@@ -129,12 +141,11 @@ jobs:
-      CC: ${{ matrix.compiler }}
-      SUFFIX: windows-${{ matrix.image }}-${{ matrix.compiler }}
+      CC: gcc
+      SUFFIX: windows-${{ matrix.image }}-gcc
```
Do `file jq.exe` for windows build also? .. sorry can't comment further down, no diff

owenthereal replied:
Yes, added.

wader commented:
A reason to wait for the docker job it make sure we notice if it fails

wader commented:
Should we try to not to mix arm64/aarch64 and x86_64/amd64? just use one?

wader commented:
ARM build for windows could be interesting but maybe for later?

wader commented:
> Should we try to not to mix arm64/aarch64 and x86_64/amd64? just use one?

Sorry now i noticed this is already done in release step 👍

itchyny commented on .github/workflows/ci.yml:
```diff
@@ -13,33 +13,34 @@ jobs:
         run: |
-          sudo apt-get update -qq
-          sudo apt-get install -y automake autoconf
+          sudo apt-get install -y automake autoconf gcc-${ARCH/_/-}-linux-gnu
```
Package name construction like this is very hard to read. Could you define explicitly in the matrix include? Also, I don't think we need this package if compiler is clang but am I correct?

nicowilliams replied:
I'm ok with this.

itchyny commented on .github/workflows/ci.yml:
```diff
@@ -258,7 +269,7 @@ jobs:
     permissions:
       contents: write
-    needs: [linux, macos, windows, dist, docker]
+    needs: [linux, macos, windows, dist]
```
We should abort releasing if docker build/publish fails. Even if this change releases the artifact quickly, this isn't good.

owenthereal replied:
To clarify, the `release` job creates a draft release instead of releasing it. However, I understand you want to wait for it before creating the draft release. I put it back, provided that you will improve the docker job speed by adopting @userdocs's suggestions.

itchyny commented on .github/workflows/ci.yml:
```diff
@@ -101,10 +109,14 @@ jobs:
             --enable-static \
-            --enable-all-static
+            --enable-all-static \
+            --host=${{ matrix.arch }}-apple-darwin$(uname -r)
```
Just nitpick but make sure to `actionlint .github/workflows/ci.yml` pass ([`actionlint`](https://github.com/rhysd/actionlint) is available by `go install github.com/rhysd/actionlint/cmd/actionlint@latest`).

owenthereal replied:
TIL. `actionlint` is pretty handy!

userdocs commented:
So I was looking at this closed pr https://github.com/jqlang/jq/pull/2652 and it seems apparent the obvious issue was that you are building via emulation, instead of cross compiling and then adding the prebuilt binary to the respective target arch, which would dramatically decrease the time to completion for that job. There is no real good reason to emulate the entire process and not on Github, you'd need to use something like https://buildjet.com/for-github-actions to make it faster.

Debian has these toolchains available https://packages.debian.org/source/bookworm/build-essential and when you use that to cross build you can do something like this perhaps?

> Note: Using a debian docker (no need for ubuntu specifically here as they get these from Debian anyway and we are building statically)

https://github.com/userdocs/jq-cross-debian/actions/runs/5487472986 to publish binaries as release assets https://github.com/userdocs/jq-cross-debian/releases/tag/master-6944d81

And then just copy those to a docker image which will provide the same outcome at in a fraction of the time.

So now you have a full suite of cross built target arch static binaries as release assets people can download and then a very simple docker build process that takes simple copies the binary from the release asset or artifacts of the job.

In my example CI you can see the file outputs per build

https://github.com/userdocs/jq-cross-debian/actions/runs/5487472986/jobs/9998971987

For example,

amd64

```
jq-amd64: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, BuildID[sha1]=6e8d99f61ae73fd3509778c6ff5a18076251c1a3, for GNU/Linux 3.2.0, stripped
```

s390x

```
jq-s390x: ELF 64-bit MSB executable, IBM S/390, version 1 (SYSV), statically linked, BuildID[sha1]=32615f38f2b33413a77bcd70432d87bbb87517c4, for GNU/Linux 3.2.0, stripped
```

Personally i'd do musl static but since you are already using a debian based solution and the binary appears to be be fullly static with no glibc dependency using [crossbuild-essential](https://packages.debian.org/bookworm/crossbuild-essential-amd64) works well, i think.

itchyny commented:
@userdocs Thank you. I'll improve the docker job to use the pre-built executables.

EDIT: I have fixed the docker job and it runs within 3 minutes to build the multi-platform image. What a big improvement.

itchyny commented on .github/workflows/ci.yml:
```diff
@@ -273,8 +284,10 @@ jobs:
+          cp artifacts/jq-linux-ubuntu-22.04-gcc-x86_64/jq release/jq-linux-amd64
+          cp artifacts/jq-linux-ubuntu-22.04-gcc-aarch64/jq release/jq-linux-arm64
+          cp artifacts/jq-macos-macos-13-clang-x86_64/jq release/jq-macos-amd64
+          cp artifacts/jq-macos-macos-13-clang-arm64/jq release/jq-macos-arm64
```
We can mv instead of cp here.

owenthereal committed:
Cross compile to more Linux platforms

owenthereal committed:
Simplify MacOS target

owenthereal committed:
Fix cp the right bin for release job

owenthereal committed:
Add missing Linux strip

owenthereal committed:
Cross compile for Windows

owenthereal committed:
Don't define pthread related functions for MinGW

MinGW includes pthread related functions. Ref: https://github.com/owenthereal/jq/actions/runs/5496335434/jobs/10016293713#step:4:1237

owenthereal committed:
Upload Windows artifacts

owenthereal committed:
Use msys gcc for Win64 build

Tests failed for mingw64 build

owenthereal committed:
Rename upload artifacts

owenthereal committed:
Remove explicit `strip`

`LDFLAGS=-s` already strip symbols

owenthereal committed:
Revert that release job waits for docker job

owenthereal committed:
`make -j$(proc)` as much as possible

owenthereal committed:
Remove unused log

owenthereal commented:
Thanks @userdocs for the awesome example of cross-compiling on Linux with [crossbuild-essential](https://packages.debian.org/bookworm/crossbuild-essential-amd64)! I adopted your approach for the Linux build in this PR. I didn't change the `docker` job yet, considering [@itchyny is going to](https://github.com/jqlang/jq/pull/2665#issuecomment-1626405755). I imagine the docker job will copy the Linux artifacts from the `linux` job to a docker image.

Besides, I added [cross-compliation for Windows](https://github.com/jqlang/jq/pull/2665#discussion_r1257425776). Now we have complete coverage of builds for most platforms 😄. The `release` job is also updated to upload all the artifacts. [Here](https://github.com/owenthereal/jq/releases/tag/jq-0.0.22) is an example.

Last but not least, we wouldn't need to explicitly `strip` and `LDFLAGS=-s` will also strip symbols.

This PR is ready to be reviewed again.

userdocs commented:
@owenthereal glad it helped. I had used LDFLAGS since the configure appears to look for a non existent strip binaries and never find it, perhaps meaning the configure needs to be overhauled at some point. With the linker flags, the gcc doing the cross compiling will strip it using strip binary it comes with, so made sense as way around that.

and also for @itchyny, As for docker, after my post, I attempted to add a matrix multi platform (which would further increase speed) job but it's not quite working. Following this https://docs.docker.com/build/ci/github-actions/multi-platform/ it seems it need to merge the manifests in a following job.

I think a combination of the prebuilt binaries + matrixed docker multi platform job would be the fastest it could be.

But the matrix part of the docker/buildx job seems to be adding complexity and not sure what the actual benefit of that is yet.

userdocs commented:
@owenthereal also, since you are using ubuntu they provide a riscv64 crossbuild that Debian does not.

https://packages.ubuntu.com/jammy/crossbuild-essential-riscv64

Since i used Debian i did not provide it but if you add it for ubuntu you used their full suite.

https://packages.ubuntu.com/search?suite=jammy&searchon=names&keywords=crossbuild-essential

```
      arch:
          [
            riscv64,
          ]
        include:
          - arch: riscv64
            CC: "riscv64-linux-gnu"
```

itchyny commented on .github/workflows/ci.yml:
```diff
@@ -265,17 +347,15 @@ jobs:
-          cp artifacts/jq-dist/jq-* release/
+          mkdir -p "release"
+          for files in jq-*; do
+              cp -rf "${files}/*" "release/"
```
Bash does not expand wildcard in `"${files}/*"`.

owenthereal replied:
Nice find. Addressed in https://github.com/jqlang/jq/pull/2665/commits/8b486ab861db64efd5f9b65e8e26a815cb72feec

owenthereal committed:
Fix quoting of *

owenthereal committed:
Add riscv64 build for Linux

owenthereal commented:
@userdocs:

> also, since you are using ubuntu they provide a riscv64 crossbuild that Debian does not.

Added [riscv64](https://github.com/jqlang/jq/pull/2665/commits/4471bf95e3c954fcba901a0c958b5eb2a55c20d2) to the list. And [build passed](https://github.com/jqlang/jq/actions/runs/5500773619/jobs/10023920230#step:4:1304) 😄

itchyny reviewed (approved):
Thank you!

wader commented:
Close #2386?
